hazelcast/src/main/java/com/hazelcast/client/impl/protocol/util/ClientMessageBuilder.java

```diff
@@ -0,0 +1,102 @@
+package com.hazelcast.client.impl.protocol.util;
+
+import com.hazelcast.client.impl.protocol.ClientMessage;
+import com.hazelcast.nio.IOService;
+import com.hazelcast.nio.tcp.TcpIpConnection;
+
+import java.nio.ByteBuffer;
+
+import static com.hazelcast.client.impl.protocol.ClientMessage.BEGIN_AND_END_FLAGS;
+import static com.hazelcast.client.impl.protocol.ClientMessage.BEGIN_FLAG;
+import static com.hazelcast.client.impl.protocol.ClientMessage.END_FLAG;
+
+/**
+ *
+ * Builds {@link ClientMessage}s from byte chunks. Fragmented messages are merged into single messages before processed.
+ *
+ */
+public class ClientMessageBuilder {
+
+    private final Int2ObjectHashMap<BufferBuilder> builderBySessionIdMap = new Int2ObjectHashMap<BufferBuilder>();
+
+    private ClientMessageAccumulator message = new ClientMessageAccumulator();
+
+    final TcpIpConnection connection;
+    final IOService ioService;
+
+    public ClientMessageBuilder(TcpIpConnection connection) {
+        this.connection = connection;
+        this.ioService = connection.getConnectionManager().getIOHandler();
+    }
+
+    public void onData(final ByteBuffer buffer) {
+        message.accumulate(buffer);
+
+        if (!message.isComplete()) {
+            return;
+        }
+
+        //MESSAGE IS COMPLETE HERE
+        final byte flags = (byte) message.getFlags();
+
+        if ((flags & BEGIN_AND_END_FLAGS) == BEGIN_AND_END_FLAGS) {
+            //HANDLE-MESSAGE
+            ClientMessage cm = new ClientMessage();
+            cm.wrapForDecode(cm.buffer().byteBuffer(), 0);
```

Comment:
What is this buffer? Where is it coming from? Maybe it just misses a comment but it doesn't look to me as it make sense :)